### ISSUE TITLE
macros generating terra functions: definition of this variable is not in scope

### ISSUE THREAD

**Issue by Lupus** [zdevito/terra#200]
Running the following snippet results in an error "definition of this variable is not in scope" (see full backtrace below):
```terra
local partial = macro(function(fn, ...)
        assert(fn.tree.type:ispointertofunction())
        local fn_type = fn.tree.type.type

        local args = {...}
        local free_args = {}
        for i=#args+1,#fn_type.parameters do
                local s = symbol(fn_type.parameters[i])
                table.insert(free_args, s)
                table.insert(args, s)
        end
        return terra([free_args])
                return fn([args])
        end
end)

terra call_fn(fn: {} -> int)
        return fn()
end

terra foo(a: int, b: int)
        return a + b
end

terra run()
        var p = partial(foo, 10)
        var p2 = partial(p, 3)
        return call_fn(p2)
end

run()
```

Full output of terra launch:
```
$ terra ./sample.t
src/terralib.lua:1114: Errors reported during typechecking.
../sample.t:27: definition of this variable is not in scope
        var p2 = partial(p, 3)
                         ^

stack traceback:
        [C]: in function 'error'
        src/terralib.lua:388: in function 'finishandabortiferrors'
        src/terralib.lua:3279: in function 'typecheck'
        src/terralib.lua:1114: in function 'userfn'
        ./sample.t:27: Errors reported during invoking macro
        var p2 = partial(p, 3)
                        ^
        src/terralib.lua:2763: in function 'docheck'
        src/terralib.lua:2985: in function 'checkexp'
        src/terralib.lua:2512: in function 'checkexpressions'
        src/terralib.lua:3152: in function 'checksingle'
        src/terralib.lua:3186: in function 'checkstmts'
        src/terralib.lua:3081: in function 'checkblock'
        src/terralib.lua:3270: in function 'typecheck'
        src/terralib.lua:1097: in function 'defineobjects'
        ./sample.t:17: in main chunk
```

I'm using terra-Linux-x86_64-332a506. Is it possible to wrap dynamically generated terra functions multiple times within one terra function? If I remove `p2`, it works just fine with single `partial()` call.

**Comment by Lupus** [comment]
With current master (65bb35b) the error is the same (line numbers in the lua backtrace are different).

```
src/terralib.lua:1126: Errors reported during typechecking.
./sample.t:27: definition of this variable is not in scope
        var p2 = partial(p, 3)
                         ^

stack traceback:
        [C]: in function 'error'
        src/terralib.lua:388: in function 'finishandabortiferrors'
        src/terralib.lua:3300: in function 'typecheck'
        src/terralib.lua:1126: in function 'userfn'
        ./sample.t:27: Errors reported during invoking macro
        var p2 = partial(p, 3)
                        ^
        src/terralib.lua:2783: in function 'docheck'
        src/terralib.lua:3005: in function 'checkexp'
        src/terralib.lua:2532: in function 'checkexpressions'
        src/terralib.lua:3173: in function 'checksingle'
        src/terralib.lua:3207: in function 'checkstmts'
        src/terralib.lua:3101: in function 'checkblock'
        src/terralib.lua:3291: in function 'typecheck'
        src/terralib.lua:1109: in function 'defineobjects'
        ./sample.t:17: in main chunk```

**Comment by Lupus** [comment]
if I change `run()` as follows:

```terra
terra run()
         var p2 = partial(partial(foo, 10), 3)
         return call_fn(p2)
 end
```

It works as expected.

**Comment by Lupus** [comment]
I minimized the test case:

```terra
local return_n = macro(function(n)
        return terra()
                return [n]
        end
end)

terra run()
        var i = 0
        var rn = return_n(i)
        return rn()
end
```

This fails with the error mentioned above. If I change `return_n(i)` to `return_n(0)`, it works.

Backtrace for the reference:

```
src/terralib.lua:1114: Errors reported during typechecking.
./sample.t:44: definition of this variable is not in scope
        var rn = return_n(i)
                          ^

stack traceback:
        [C]: in function 'error'
        src/terralib.lua:388: in function 'finishandabortiferrors'
        src/terralib.lua:3279: in function 'typecheck'
        src/terralib.lua:1114: in function 'userfn'
        ./sample.t:44: Errors reported during invoking macro
        var rn = return_n(i)
                         ^
        src/terralib.lua:2763: in function 'docheck'
        src/terralib.lua:2985: in function 'checkexp'
        src/terralib.lua:2512: in function 'checkexpressions'
        src/terralib.lua:3152: in function 'checksingle'
        src/terralib.lua:3186: in function 'checkstmts'
        src/terralib.lua:3081: in function 'checkblock'
        src/terralib.lua:3270: in function 'typecheck'
        src/terralib.lua:1097: in function 'defineobjects'
        ./sample.t:42: in main chunk
```

**Comment by zdevito** [comment]
Using the macro to generate a new terra function in the way you are calling
it is equivalent to writing this code directly:

terra anon()
    return i
end

terra run()
        var i = 0
        var rn = anon
        return rn()
end


The problem is that the variable `i` is not actually defined in the
anonymous function. Fixing this depends on what you are trying to
accomplish in the non-reducted example, which I can help with if you want
to describe it in more detail.

**Comment by Lupus** [comment]
Yeah, I started to suspect that after numerous attempts to trick it :) Thanks for the explanation!

My original goal is to create a terra library wrapping libev, which is an event loop implementation in C. libev has a concept of watchers (i/o, timers, etc.), which have opaque `void *data` and associated callback, which typically looks like this:

```C
static void my_cb (struct ev_loop *loop, ev_io *w, int revents)
{
    //get w->data, cast to appropriate structure and use
}
```

Obviously it can be used as is. Just pass plain function and context pointer, and cast it to the appropriate type in terra function. 

In Go it is possible to pass object.method as callback and this construct will implicitly capture this/self reference. This looks much like partial function application, and as terra provides instruments instead of policies, I decided to implement partial function application as a general utility function. They I would provide the callback to the watcher in the form of partial application of the desired method along with self pointer.

As of the partial application, I've managed to implement it in somewhat generic way, but it looks messy as it involves dynamic structs, global llvm variables with pointers to those structs and dynamic functions using those pointers. Dynamic structs are heap allocated and thus they need explicit bookkeeping, and as they are dynamic, it's not easy to keep them in some statically defined structure as their type is not known. This approach seems somewhat wrong to me, the result is not so seamless.

Another idea I have is to leverage interfaces from test/lib/golike.t and add a method to libev watcher wrapper object to set event listener, which would be an interface, and clients will just make sure they have a set of functions implementing the interface.
